src/inbox/mod_inbox.erl

```diff
@@ -193,15 +212,114 @@ build_result_iq(CountBin) ->
     #xmlel{name = <<"count">>, attrs = [{<<"xmlns">>, ?NS_ESL_INBOX}],
         children = [#xmlcdata{content = CountBin}]}.
 
--spec build_forward_el(content()) -> exml:element().
-build_forward_el(Content) ->
+-spec build_forward_el(content(), erlang:timestamp()) -> exml:element().
+build_forward_el(Content, Timestamp) ->
     {ok, Parsed} = exml:parse(Content),
-    #xmlel{name = <<"forwarded">>, attrs = [{<<"xmlns">>, ?NS_FORWARD}], children = [Parsed]}.
+    Delay = build_delay_el(Timestamp),
+    #xmlel{name = <<"forwarded">>, attrs = [{<<"xmlns">>, ?NS_FORWARD}],
+           children = [Delay, Parsed]}.
+
+-spec build_delay_el(Timestamp :: erlang:timestamp()) -> exml:element().
+build_delay_el({_, _, Micro} = Timestamp) ->
+    {Day, {H, M, S}} = calendar:now_to_datetime(Timestamp),
+    DateTimeMicro = {Day, {H, M, S, Micro}},
+    jlib:timestamp_to_xml(DateTimeMicro, utc, undefined, undefined).
+
+-spec build_inbox_form() -> exml:element().
+build_inbox_form() ->
+    OrderOptions = [
+                    {<<"Ascending by timestamp">>, <<"asc">>},
+                    {<<"Descending by timestamp">>, <<"desc">>}
+                   ],
+    FormFields = [
+                  jlib:form_field({<<"FORM_TYPE">>, <<"hidden">>, ?NS_ESL_INBOX}),
+                  text_single_form_field(<<"start">>),
+                  text_single_form_field(<<"end">>),
+                  list_single_form_field(<<"order">>, <<"desc">>, OrderOptions)
+                 ],
+    #xmlel{ name = <<"x">>, attrs = [{<<"xmlns">>, ?NS_XDATA}, {<<"type">>, <<"form">>}],
+            children = FormFields }.
+
+-spec text_single_form_field(Var :: binary()) -> exml:element().
```

Comment:
I wonder if this kind of element builders are used in other places of MongooseIM. If so I hope they could be re-used. If not I think it'd be worth to extract `text_single_form_field` and other such builder function to a dedicated module (`mongoose_xmpp`?).

Reply:
Will be probably covered by a separate XMPP library.